tests/DelegationSpec.hs

```diff
@@ -0,0 +1,170 @@
+    }
+
+
+spec :: Spec
+spec = {- tag Large . -} do
+    runner   <- runIO createActionRunner
+    persist  <- runIO Persistent.mkRunPersistInMemory
+    uni      <- runIO $ unNat (runner persist) (mkUniverse universeSize)
+    snapshot <- runIO $ unNat (runner persist) getDBSnapShot
+    let programGen = delegationProgram
+                        (QC.elements $ unStudents uni)
+                        (QC.elements $ unIdeas    uni)
+                        (QC.elements $ universeToDelegationContexts uni)
+    let student1 = unStudents uni !! 1
+        student2 = unStudents uni !! 2
+        idea     = unIdeas    uni !! 1
+    describe "Delegation simulation" $ do
+        it "One delegation, one vote" $ do
+            persist' <- Persistent.mkRunPersistInMemoryWithState snapshot
+            unNat (runner persist') . interpretDelegationProgram $ DelegationProgram
+                [ SetDelegation student1 (DlgCtxIdeaId idea) student2
+                , Vote student1 idea Yes
+                ]
+        it "Circle in delegation" $ do
+            persist' <- Persistent.mkRunPersistInMemoryWithState snapshot
+            unNat (runner persist') . interpretDelegationProgram $ DelegationProgram
+                [ SetDelegation student1 (DlgCtxIdeaId idea) student2
+                , CheckNoOfDelegatees student2 (DlgCtxIdeaId idea) 2
+                , SetDelegation student2 (DlgCtxIdeaId idea) student1
+                , CheckNoOfDelegatees student1 (DlgCtxIdeaId idea) 2
+                , Vote student1 idea Yes
+                , Vote student2 idea Yes
+                ]
+        it "Random delegation programs" . property . forAllShrinkDef programGen $ \prg -> do
+            monadicIO $ do
+                persist' <- run $ Persistent.mkRunPersistInMemoryWithState snapshot
+                run . unNat (runner persist') $ interpretDelegationProgram prg
+  where
+    getDBSnapShot :: Action.Action Persistent.AulaData
+    getDBSnapShot = query (view Persistent.dbSnapshot)
+
+    createActionRunner :: IO (Persistent.RunPersist -> (Action.Action :~> IO))
+    createActionRunner = do
+        cfg <- testConfig
+        let runAction :: Persistent.RunPersist -> (Action.Action :~> IO)
+            runAction persist = exceptToFail . Action.mkRunAction (Action.ActionEnv persist cfg nullLog)
+
+        return runAction
+
+    universeToDelegationContexts :: Universe -> [DelegationContext]
+    universeToDelegationContexts u = DlgCtxGlobal:(spaces <> topics <> ideas)
+      where
+        spaces = DlgCtxIdeaSpace <$> unIdeaSpaces u
+        topics = DlgCtxTopicId   <$> unTopics     u
+        ideas  = DlgCtxIdeaId    <$> unIdeas      u
+
+-- * delegation program
+
+data DelegationDSL where
+    SetDelegation       :: AUID User -> DelegationContext -> AUID User      -> DelegationDSL
+    Vote                :: AUID User -> AUID Idea         -> IdeaVoteValue  -> DelegationDSL
+    CheckNoOfDelegatees :: AUID User -> DelegationContext -> Int -> DelegationDSL
```

Comment:
rename to `CheckNumVotingPower`.  it's number of delegatees (aka delegators) plus oneself.